opm/simulators/wells/WellGroupHelpers.hpp

```diff
@@ -0,0 +1,211 @@
+
+            int well_index = it->second[0];
+
+            const auto& wellEcl = schedule.getWell2(wellName, reportStepIdx);
+            //only count producers or injectors
+            if ( (wellEcl.isProducer() && injector) ||  (wellEcl.isInjector() && !injector))
+                continue;
+
+            double factor = wellEcl.getEfficiencyFactor();
+
+            const auto wellrate_index = well_index * wellState.numPhases();
+            if (injector)
+                rate += factor * wellState.wellReservoirRates()[ wellrate_index + phasePos];
+            else
+                rate -= factor * wellState.wellReservoirRates()[ wellrate_index + phasePos];
+        }
+    }
+
+    inline void setGroupControl(const Group2& group, const Schedule& schedule, const int reportStepIdx, const bool injector, WellStateFullyImplicitBlackoil& wellState) {
+
+        for (const std::string& groupName : group.groups()) {
+            const Group2& groupTmp = schedule.getGroup2(groupName, reportStepIdx);
+            setGroupControl(groupTmp, schedule, reportStepIdx, injector, wellState);
+            if (injector)
+                wellState.setCurrentInjectionGroupControl(groupName, "FLD");
+            else
+                wellState.setCurrentProductionGroupControl(groupName, "FLD");
+        }
+
+        const auto& end = wellState.wellMap().end();
+        for (const std::string& wellName : group.wells()) {
+            const auto& it = wellState.wellMap().find( wellName );
+            if (it == end)  // the well is not found
+                continue;
+
+            int well_index = it->second[0];
+            const auto& wellEcl = schedule.getWell2(wellName, reportStepIdx);
+            if (!wellEcl.isAvailableForGroupControl())
+                continue;
+            if (wellEcl.isProducer() && !injector)
+                wellState.currentControls2()[well_index] = WellProducer::ControlMode2String(WellProducer::GRUP);
+
+            if (wellEcl.isInjector() && injector)
+                wellState.currentControls2()[well_index] = WellInjector::ControlMode2String(WellInjector::GRUP);
+        }
+    }
+
+    inline double computeGuideRate(const WellStateFullyImplicitBlackoil& wellState, const Well2& well, const int phasePos )
+    {
+        const auto& end = wellState.wellMap().end();
+        const auto& it = wellState.wellMap().find( well.name() );
+        if (it == end)  // the well is not found
+            return 0.0;
+
+        int well_index = it->second[0];
+
+        if (wellState.currentControls2()[well_index] !="GRUP")
+            return 0.0;
+
+        GuideRate::GuideRatePhaseEnum GuideRatePhase = well.getGuideRatePhase();
+#warning handle the case when phase != GuideRatePhase
+        if (GuideRatePhase == Opm::GuideRate::UNDEFINED) {
+            const auto wellrate_index = well_index * wellState.numPhases();
+            double guideRate = wellState.wellPotentials()[wellrate_index + phasePos];
+            if (guideRate == 0)
```

Comment:
This is wrong. Deleting these two lines makes the SPE9_CP_GROUP case almost as good as master. There are still a few spikes just when switching to group control though.

If there is a possibility of the potential not being updated correctly, we should rather fix that than include this line. If we need to safeguard against total guide rates being zero that should happen where we use it.

Reply:
You are right. This is a unfortunate hack to avoid division by zero. I should have added a #warning. The reason for this is partly in the initialization of the well potential, but also that we return zero potential if the potential calculations failed. I didn't now it affected SPE9_CP_GROUP, thanks for noticing. I have warried a bit about the "spikes" on SPE9_CP_GROUP. I assumed it was related to the initialization or the setWellStateFromTarget, but I haven't fixed it.